enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,327 @@
+- (This doesn’t benefit from service-ca involvement at all.)
+
+Similarly single-purpose CA could be set up for NodeIP services
+(if it makes sense to connect to node IP addresses directly at all)
+so that the client connects to the desired NodeIP service
+instead of to any NodeIP service.
+
+### Risks and Mitigations
+
+The only mitigations for the risks below are not to use this feature,
+(and prevent untrusted users from using it — using an admission controller?)
+so that the wildcard certificate is not created at all,
+or perhaps $somehow ensure that any users/code with access to the wildcard certificate
+is not likely to be compromised.
+
+#### Unexpected DNS Collisions in the Future
+
+https://github.com/kubernetes/dns/blob/master/docs/specification.md
+only specifies what DNS records must exist, not what must not exist,
+so the generated wildcard certificate could match unwanted DNS entries.
+
+The DNS specification does not even guarantee that different namespaces
+will use different DNS subtrees.
+
+Given the existing naming conventions of in-cluster DNS,
+such unexpected collisions seem somewhat unlikely,
+but can’t be ruled out in advance.
+
+#### No Individual Pod Identity
+
+The wildcard certificate does not protect the intra-StatefulSet communication
+against re-routing to unintended members of that StatefulSet
+(or to a different StatefulSet attached to the same headless service).
+
+In addition, any application that uses a different set of certificates to achieve that goal
+**must not accept the Service CA as a valid root of trust for connections intended
+for individual StatefulSet pods**.
+
+## Design Details
+
+### Open Questions
+
+N/A
+
+### Test Plan
+
+Unit and e2e tests similar to the existing service certificate tests, both in structure and scope.
+
+Ensure that upgrades (or alternatively any controller re-syncs)
+cause re-generation of certificates for headless services
+to include the wildcard subjects, if they are missing.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation (TBD: where?)
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+  TBD?
+- Write symptoms-based alerts for the component(s)
+  TBD?
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+  TBD?
+- Document SLOs for the component
+  TBD?
+- Conduct load testing
+  TBD?
+
+### Upgrade / Downgrade Strategy
+
+Upgrading to a version that adds the feature will trigger a regeneration of secrets
+requested for headless services.
+The previously-generated certificates will continue to be valid for their original lifetime,
+and previously-deployed components can continue to use them as before.
+
+If the upgrade to the first version that generates the wildcards
```

Comment:
I do not believe there are any operators relying on the combination of statefulsets+service ca today, and that is likely to remain true until the proposed feature lands in 4.8. As downgrades are only supported for emergency scenarios caused by failed upgrades (and in the CI testing that validates those scenarios), once a cluster is successfully upgraded to 4.8 any component would be able to safely rely on the proposed feature being available.

Reply:
Thanks, simplified.